File: YJ/20200904/BinSearch/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int num_of_rocks(int dist, vector<int> rocks) {
    int now_point = 0, cnt = 0;
    for (int i = 0; i < rocks.size(); i ++) {
        if (rocks[i] - now_point >= dist) {
            now_point = rocks[i];
            cnt ++;
        }
    }
    return cnt;
}

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0, lft, rt, mid;
    
    sort(rocks.begin(), rocks.end());
    
    lft = 1;
    rt = distance;
    
    while(lft <= rt) {
        mid = (lft + rt) / 2;
        
        if (num_of_rocks(mid, rocks) >= rocks.size() - n) {
            lft = mid + 1;
            answer = mid;
        } else {
            rt = mid - 1;
        }
    }
    
    return answer;
}

File: JY/8/8.28/징검다리.py
def solution(distance, rocks, n):
    ans = 0
    rocks.sort()
    rn = len(rocks)
    start = 0
    end = distance
    while start<end:
        mid = (start+end)//2
        cnt = 0
        tmp = 0
        flag = False
        for i in range(rn):
            if rocks[i]-tmp<mid:
                if cnt<n:
                    cnt += 1
                    continue
                else:
                    flag = True
                    break
            else:
                tmp = rocks[i]
        print(mid)   
        if flag:
            end = mid
        else:
            start = mid +1
    ans = end-1
    return ans